File: go/src/337.打家劫舍-iii.go
/*
 * @lc app=leetcode.cn id=337 lang=golang
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rob(root *TreeNode) int {
	counter := map[*TreeNode]int{}
	return util(root, counter)
}

func util(root *TreeNode, counter map[*TreeNode]int) int {
	if root == nil {
		return 0
	}

	if v, ok := counter[root]; ok {
		return v
	}

	res1 := root.Val
	if root.Left != nil {
		res1 += util(root.Left.Left, counter) + util(root.Left.Right, counter)
	}
	if root.Right != nil {
		res1 += util(root.Right.Left, counter) + util(root.Right.Right, counter)
	}

	res2 := util(root.Left, counter) + util(root.Right, counter)
	counter[root] = max(res1, res2)
	return counter[root]
}

func max(a ...int) int {
	res := a[0]
	for _, v := range a {
		if res < v {
			res = v
		}
	}
	return res
}
// @lc code=end

File: rust/src/337.打家劫舍-iii.rs
/*
 * @lc app=leetcode.cn id=337 lang=rust
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn rob(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::dfs(&root)[1]
    }

    // res[0] 代表不抢劫root的结果   res[1] 代表最优值
    fn dfs(root: &Option<Rc<RefCell<TreeNode>>>) -> [i32; 2] {
        if let Some(root) = root {
            let left = Self::dfs(&root.borrow().left);
            let right = Self::dfs(&root.borrow().right);

            let val = left[0] + right[0] + root.borrow().val;
            let val2 = left[1] + right[1];
            [val2, val.max(val2)]
        } else {
            [0; 2]
        }
    }
}
// @lc code=end
